_posts/2021-08-16-connector-table-sql-api-part1.md

```diff
@@ -0,0 +1,234 @@
+
+```java
+import java.util.HashSet;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.factories.DynamicTableSourceFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+
+public class ImapTableSourceFactory implements DynamicTableSourceFactory {
+  @Override
+  public String factoryIdentifier() {
+    return "imap";
+  }
+
+  @Override
+  public Set<ConfigOption<?>> requiredOptions() {
+    return new HashSet<>();
+  }
+
+  @Override
+  public Set<ConfigOption<?>> optionalOptions() {
+    return new HashSet<>();
+  }
+
+  @Override
+  public DynamicTableSource createDynamicTableSource(Context ctx) {
+    final FactoryUtil.TableFactoryHelper factoryHelper = FactoryUtil.createTableFactoryHelper(this, ctx);
+    factoryHelper.validate();
+
+    return new ImapTableSource();
+  }
+}
+```
+
+There are currently no configuration options but they can be added and also validated within the `createDynamicTableSource()` function. There is a small helper utility, [TableFactoryHelper](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/FactoryUtil.TableFactoryHelper.html), that Flink offers which ensures that required options are set and that no unknown options are provided.
+
+Finally, you need to register your factory for Java's Service Provider Interfaces (SPI). Classes that implement this interface can be discovered and should be added to this file `src/main/resources/META-INF/services/org.apache.flink.table.factories.Factory` with the fully classified class name of your factory:
+
+```java
+// if you created your class in the package org.example.acme, it should be named the following:
+org.example.acme.ImapTableSourceFactory
+```
+
+# Test the custom connector
+
+You should now have a working source connector. If you are following along with the provided repository, you can test it by running:
+
+```sh
+$ cd testing/
+$ ./build_and_run.sh
+```
+
+This builds the connector, starts a Flink cluster, a [test email server](https://greenmail-mail-test.github.io/greenmail/) which you will need later), and the [SQL client](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/table/sqlclient/) for you. You may need to run this script as a super user. If successful, you should see the SQL CLI:
```

Comment:
> You may need to run this script as a super user.

I would refrain from this piece of advice in a blog post. Running docker containers as root is a bad idea. Yes, people can get stuck because they have no proper docker setup on their machine, and we don't have the time in a blog post to explain all of that, but teaching people to give random scripts or docker containers root access is wrong.

I could help here on the script side by catching this case and printing a helpful error message instead.